pkg/blockbuilder/blockbuilder.go

```diff
@@ -358,47 +426,38 @@ func (b *BlockBuilder) consumePartitions(ctx context.Context, cycleEnd time.Time
 				break
 			}
 
-			// TODO(v): refactor block bounds check into a state machine:
-			// During catching up, if builder skipped samples, roll back the offset to first uncommited after compacting and starting next block
-			if rec.Timestamp.Before(blockStartAt) || rec.Timestamp.After(blockEndAt) {
-				// When BB is first deployed or if it is lagging behind, then it might consuming data from too much
-				// in the past. In which case if we try to consume all at once, it can overwhelm the system.
-				// So we break this into multiple block building cycles.
-				if rec.Timestamp.After(blockEndAt) {
-					if err := builder.compactAndUpload(ctx, b.blockUploaderForUser); err != nil {
-						return fmt.Errorf("compact tsdb builder: %w", err)
-					}
-				}
-				blockStartAt, blockEndAt = blockBounds(rec.Timestamp, b.cfg.ConsumeInterval)
-			}
+			lastRec = rec
 
-			level.Debug(b.logger).Log("msg", "process record", "offset", rec.Offset, "rec", rec.Timestamp, "bmin", blockStartAt, "bmax", blockEndAt)
+			level.Debug(b.logger).Log("msg", "process record", "offset", rec.Offset, "rec", rec.Timestamp, "bmin", lastBlockEnd, "bmax", blockEndAt)
 
-			recProcessedBefore := false // TODO: get this from kafka commit
-			blockMin, blockMax := blockStartAt.UnixMilli(), blockEndAt.UnixMilli()
-			allSamplesProcessed, err := builder.process(ctx, rec, blockMin, blockMax, recProcessedBefore)
+			recProcessedBefore := seenTillTs.After(rec.Timestamp)
+			allSamplesProcessed, err := builder.process(ctx, rec, lastBlockMax, blockMax, recProcessedBefore)
 			if err != nil {
 				level.Error(b.logger).Log("msg", "failed to process record", "part", part, "key", string(rec.Key), "err", err)
 				// TODO(codesome): do we just ignore this? What if it was Mimir's issue and this leading to data loss?
 				// TODO(codesome): add metric
-			} else if allSamplesProcessed {
-				// TODO(v): only advance checkpoint when all previously seen records from this cycle were processed
+			}
+			if !allSamplesProcessed && checkpointRec == nil {
+				// First record where all samples were not processed becomes the checkpoint/commit point.
 				checkpointRec = rec
```

Comment:
> First record where all samples were not processed becomes the checkpoint/commit point

I think this isn't right, 'cause checkpoint must be at the offset **before** this record. That is, in Kafka commit is the offset of that was processed. The next consumption starts from the offset after.

Consider this log

```
{offset=1}, {offset=2}, {offset=3}, ...
```

The happy case of the consumer is the following

1. consume the next batch from broker, e.g. a fetch `{offset=1}, {offset=2}`
2. process the records in the batch
3. commit the offset of the last processed record in the batch, `offset=2` for this example
4. go to 1

On restart, the consumer looks up the last committed offset of its group (here, `offset=2`), and starts consuming from the next offset.

Reply:
So a off by 1 error. I will fix it. Thanks!